--- a/drivers/net/ethernet/realtek/r8169soc.c
+++ b/drivers/net/ethernet/realtek/r8169soc.c
@@ -128,7 +128,7 @@ enum cfg_version {
 static int rx_buf_sz = 1523;	/* 0x05F3 = 1522bye + 1 */
 static int rx_buf_sz_new = 1523;
 #else
-static int rx_buf_sz = 1523;
+static int rx_buf_sz = 16383;
 #endif /* CONFIG_RTL_RX_NO_COPY */
 
 static struct {
@@ -3235,9 +3235,6 @@ static void rtl_hw_start_8168(struct net
 
 static int rtl8169_change_mtu(struct net_device *dev, int new_mtu)
 {
-	if (new_mtu > ETH_DATA_LEN)
-		new_mtu = ETH_DATA_LEN;
-
 #if defined(CONFIG_RTL_RX_NO_COPY)
 	struct rtl8169_private *tp = netdev_priv(dev);
 
