From 255e5ccca526b856baf05a413e4f6d3a73dfcc50 Mon Sep 17 00:00:00 2001
From: KenYen <KenYen@realtek.com>
Date: Wed, 9 Apr 2025 18:25:41 +0800
Subject: recipes-kernel: linux: r8169soc: 704: support jumbo frame
Source: Realtek
License: GPL-2.0

Change-Id: I94cdf9efe7cb29f31a1556d0a41a5753c8a744f5
---
--- a/drivers/net/ethernet/realtek/r8169soc.c
+++ b/drivers/net/ethernet/realtek/r8169soc.c
@@ -128,7 +128,7 @@ enum cfg_version {
 static int rx_buf_sz = 1523;	/* 0x05F3 = 1522bye + 1 */
 static int rx_buf_sz_new = 1523;
 #else
-static int rx_buf_sz = 1523;
+static int rx_buf_sz = 16383;
 #endif /* CONFIG_RTL_RX_NO_COPY */
 
 static struct {
@@ -3235,9 +3235,6 @@ static void rtl_hw_start_8168(struct net
 
 static int rtl8169_change_mtu(struct net_device *dev, int new_mtu)
 {
-	if (new_mtu > ETH_DATA_LEN)
-		new_mtu = ETH_DATA_LEN;
-
 #if defined(CONFIG_RTL_RX_NO_COPY)
 	struct rtl8169_private *tp = netdev_priv(dev);
 
