From 71ef8c98d022b94b8ff5b3bcb229d554286fca26 Mon Sep 17 00:00:00 2001
From: Phinex Hung <phinex@realtek.com>
Date: Wed, 3 Jan 2024 20:08:29 +0800
Subject: [PATCH] u-boot: add u-boot patch and drivers from openwrt 22.03

---
 drivers/spi/Kconfig  | 6 ++++++
 drivers/spi/Makefile | 1 +
 2 files changed, 7 insertions(+)

diff --git a/drivers/spi/Kconfig b/drivers/spi/Kconfig
index 69b184b0d9..2382230937 100644
--- a/drivers/spi/Kconfig
+++ b/drivers/spi/Kconfig
@@ -422,6 +422,24 @@ config ROCKCHIP_SPI
 	  This uses driver model and requires a device tree binding to
 	  operate.
 
+config SPI_RTK_SFC
+	bool "Realtek SPI-NOR Flash Controller(SFC)"
+	depends on SPI_FLASH || DM_SPI_FLASH
+	help
+	  This enables support for Realtek SPI-NOR flash controller.
+
+config SPI_RTK_NAND
+	bool "RTK SPI NAND controller"
+	depends on MTD_NAND_ECC && MTD_SPI_NAND
+	help
+	  Enables SPI Nand controller driver for RTK SoC.
+
+config SPL_SPI_RTK_NAND
+	bool "RTK SPI NAND controller in SPL"
+	depends on MTD_NAND_ECC && MTD_SPI_NAND && SPL
+	help
+	  Enables SPI Nand controller driver for RTK SoC.
+
 config SANDBOX_SPI
 	bool "Sandbox SPI driver"
 	depends on SANDBOX && DM
diff --git a/drivers/spi/Makefile b/drivers/spi/Makefile
index 14bdb97f18..a34a03aa05 100644
--- a/drivers/spi/Makefile
+++ b/drivers/spi/Makefile
@@ -67,6 +67,8 @@ obj-$(CONFIG_SPI_MXIC) += spi-mxic.o
 obj-$(CONFIG_RENESAS_RPC_SPI) += renesas_rpc_spi.o
 obj-$(CONFIG_ROCKCHIP_SFC) += rockchip_sfc.o
 obj-$(CONFIG_ROCKCHIP_SPI) += rk_spi.o
+obj-$(CONFIG_SPI_RTK_SFC) += rtk-sfc.o
+obj-$(CONFIG_$(SPL_)SPI_RTK_NAND) += spi-rtk-sn.o
 obj-$(CONFIG_SANDBOX_SPI) += sandbox_spi.o
 obj-$(CONFIG_SPI_SIFIVE) += spi-sifive.o
 obj-$(CONFIG_SPI_SN_F_OSPI) += spi-sn-f-ospi.o
-- 
2.47.0

