From aac50160bc31dcde1e9e0029f5538ff451f9f5ce Mon Sep 17 00:00:00 2001
From: "joshua.yang" <joshua.yang@realtek.com>
Date: Mon, 17 Mar 2025 18:58:16 +0800
Subject: [PATCH 1/1] Support video/x-wmv with both WVC1 and WMV3 formats
Upstream-Status: Inappropriate [rtk specific]

---
 sys/v4l2/gstv4l2object.c | 33 +++++++++++++++++++++++++++++++--
 1 file changed, 31 insertions(+), 2 deletions(-)

diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index f063324..c711c00 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -1526,8 +1526,30 @@ gst_v4l2_object_v4l2fourcc_to_bare_struct (guint32 fourcc)
       break;
     case V4L2_PIX_FMT_VC1_ANNEX_G:
     case V4L2_PIX_FMT_VC1_ANNEX_L:
-      structure = gst_structure_new ("video/x-wmv",
-          "wmvversion", G_TYPE_INT, 3, "format", G_TYPE_STRING, "WVC1", NULL);
+      {
+        structure = gst_structure_new ("video/x-wmv",
+            "wmvversion", G_TYPE_INT, 3,
+            "stream-format", G_TYPE_STRING, "frame-layer",
+            "header-format", G_TYPE_STRING, "sequence-layer", NULL);
+
+        GValue list = G_VALUE_INIT;
+        g_value_init(&list, GST_TYPE_LIST);
+
+        GValue format1 = G_VALUE_INIT;
+        g_value_init(&format1, G_TYPE_STRING);
+        g_value_set_string(&format1, "WVC1");
+        gst_value_list_append_value(&list, &format1);
+
+        GValue format2 = G_VALUE_INIT;
+        g_value_init(&format2, G_TYPE_STRING);
+        g_value_set_string(&format2, "WMV3");
+        gst_value_list_append_value(&list, &format2);
+
+        gst_structure_set_value(structure, "format", &list);
+        g_value_unset(&format1);
+        g_value_unset(&format2);
+        g_value_unset(&list);
+      }
       break;
     case V4L2_PIX_FMT_VP8:
       structure = gst_structure_new_empty ("video/x-vp8");
@@ -2007,6 +2029,13 @@ gst_v4l2_object_get_caps_info (GstV4l2Object * v4l2object, GstCaps * caps,
         fourcc = V4L2_PIX_FMT_H264_NO_SC;
       else
         fourcc = V4L2_PIX_FMT_H264;
+    } else if (g_str_equal (mimetype, "video/x-wmv")) {
+      const gchar *stream_format =
+          gst_structure_get_string (structure, "format");
+      if (g_str_equal (stream_format, "WVC1"))
+        fourcc = V4L2_PIX_FMT_VC1_ANNEX_G;
+      else
+        fourcc = V4L2_PIX_FMT_VC1_ANNEX_L;
     } else if (g_str_equal (mimetype, "video/x-h265")) {
       fourcc = V4L2_PIX_FMT_HEVC;
     } else if (g_str_equal (mimetype, "video/x-vp8")) {
-- 
2.34.1

