From 0ca3308edefa3e23274b9614b74722bf0133392c Mon Sep 17 00:00:00 2001
From: "joshua.yang" <joshua.yang@realtek.com>
Date: Mon, 27 Oct 2025 19:29:07 +0800
Subject: [PATCH 1/1] Fix EOS flush running during resolution change event
 dispatch
Upstream-Status: Inappropriate [rtk specific]

---
 sys/v4l2/gstv4l2videodec.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index 6138b24..805e5ec 100644
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -878,6 +878,7 @@ gst_v4l2_video_dec_loop (GstVideoDecoder * decoder)
   GstVideoCodecFrame *frame;
   GstBuffer *buffer = NULL;
   GstFlowReturn ret;
+  gboolean should_stop_v4l2capture = FALSE;
 
   GST_LOG_OBJECT (self, "Looping.");
 
@@ -1056,9 +1057,8 @@ beach:
     GST_VIDEO_DECODER_STREAM_LOCK (decoder);
     if (self->draining) {
       self->draining = FALSE;
-      gst_v4l2_object_stop (self->v4l2capture);
+      should_stop_v4l2capture = TRUE;
       GST_VIDEO_DECODER_STREAM_UNLOCK (decoder);
-      return;
     }
 
     GST_VIDEO_DECODER_STREAM_UNLOCK (decoder);
@@ -1071,6 +1071,11 @@ beach:
   self->output_flow = ret;
   gst_v4l2_object_unlock (self->v4l2output);
   gst_pad_pause_task (decoder->srcpad);
+
+  if (should_stop_v4l2capture) {
+    GST_INFO_OBJECT (decoder, "Stopping v4l2capture after pausing task");
+    gst_v4l2_object_stop (self->v4l2capture);
+  }
 }
 
 static GstFlowReturn
-- 
2.34.1

