From 742ebf1e4c00b8d9ae4de42b3dd690684362f97b Mon Sep 17 00:00:00 2001
From: edwardwu <edwardwu@realtek.com>
Date: Fri, 2 Jul 2021 11:00:48 +0800
Subject: [PATCH] [DEV_CHG] Always inline to improve network performance

---
diff --git a/include/linux/compiler_types.h b/include/linux/compiler_types.h
index ac3fa37a8..5aacf66d4 100644
--- a/include/linux/compiler_types.h
+++ b/include/linux/compiler_types.h
@@ -134,14 +134,8 @@ struct ftrace_likely_data {
 
 #define __compiler_offsetof(a, b)	__builtin_offsetof(a, b)
 
-/*
- * Prefer gnu_inline, so that extern inline functions do not emit an
- * externally visible function. This makes extern inline behave as per gnu89
- * semantics rather than c99. This prevents multiple symbol definition errors
- * of extern inline functions at link time.
- * A lot of inline functions can cause havoc with function tracing.
- */
-#define inline inline __gnu_inline __inline_maybe_unused notrace
+/* Always inline to improve network performance */
+#define inline inline __attribute__((__always_inline__)) __gnu_inline __inline_maybe_unused notrace
 
 /*
  * gcc provides both __inline__ and __inline as alternate spellings of
