From bcb85c4caebb3441571f35de87e09cab6bd9d212 Mon Sep 17 00:00:00 2001
From: "joshua.yang" <joshua.yang@realtek.com>
Date: Wed, 23 Jul 2025 16:34:19 +0800
Subject: [PATCH 1/1] NASPRJ-1171: Add NV24 to support mjpeg render

Upstream-Status: Inappropriate [rtk specific]
---
 libweston/renderer-gl/gl-renderer.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/libweston/renderer-gl/gl-renderer.c b/libweston/renderer-gl/gl-renderer.c
index fbd96a6..dee9de3 100644
--- a/libweston/renderer-gl/gl-renderer.c
+++ b/libweston/renderer-gl/gl-renderer.c
@@ -292,6 +292,17 @@ struct yuv_format_descriptor yuv_formats[] = {
             .format = DRM_FORMAT_GR88,
             .plane_index = 1
         }}
+	}, {
+		.format = DRM_FORMAT_NV24,
+		.output_planes = 2,
+		.shader_variant = SHADER_VARIANT_Y_UV,
+		{{
+			.format = DRM_FORMAT_R8,
+			.plane_index = 0
+		}, {
+			.format = DRM_FORMAT_GR88,
+			.plane_index = 1
+		}}
 	}, {
 		.format = DRM_FORMAT_P010,
 		.output_planes = 2,
@@ -4049,6 +4060,7 @@ gl_renderer_display_create(struct weston_compositor *ec,
 	wl_display_add_shm_format(ec->wl_display, WL_SHM_FORMAT_YUV444);
 	wl_display_add_shm_format(ec->wl_display, WL_SHM_FORMAT_NV12);
 	wl_display_add_shm_format(ec->wl_display, WL_SHM_FORMAT_NV16);
+	wl_display_add_shm_format(ec->wl_display, WL_SHM_FORMAT_NV24);
 	wl_display_add_shm_format(ec->wl_display, WL_SHM_FORMAT_RGB888);
 	wl_display_add_shm_format(ec->wl_display, WL_SHM_FORMAT_YUYV);
 	wl_display_add_shm_format(ec->wl_display, WL_SHM_FORMAT_XYUV8888);
-- 
2.34.1

