From ac5d2cc908cf9a768a878efe0267bb9a836eb792 Mon Sep 17 00:00:00 2001
From: Ken Tseng <ken.tseng@realtek.com>
Date: Fri, 9 Jan 2026 11:33:50 +0800
Subject: [PATCH] Fix attribute-warning on write overflow

---
 capture.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/capture.c b/capture.c
index 3b615d8..efe5fb2 100644
--- a/capture.c
+++ b/capture.c
@@ -366,7 +366,7 @@ static int capture_s_parm(struct v4l2_fh *fh, struct v4l2_streamparm *a)
 	mutex_lock(&ctx->capture_mutex);
 
 	tpf = &a->parm.output.timeperframe;
-	memcpy(&ctx->timeperframe, tpf, sizeof(struct v4l2_format));
+	memcpy(&ctx->timeperframe, tpf, sizeof(struct v4l2_fract));
 
 	mutex_unlock(&ctx->capture_mutex);
 
@@ -813,7 +813,7 @@ static int _capture_get_wb_frame(struct capture_ringbuffer_info *rb_info ,YUV_ST
 	
 	ret = read_inband_cmd(rb_info, (uint8_t *)frame, hdr.size - sizeof(INBAND_CMD_PKT_HEADER));
     if(hdr.type != VIDEO_VO_INBAND_CMD_TYPE_OBJ_PIC || ret != 0 ) {
-        pr_info("hdr.type = %d need remain size ::%u  ",hdr.type, hdr.size - sizeof(INBAND_CMD_PKT_HEADER));
+        pr_info("hdr.type = %d need remain size ::%u  ",(int)hdr.type, (unsigned int)(hdr.size - sizeof(INBAND_CMD_PKT_HEADER)));
 		return -1;
     }
 
-- 
2.34.1

