diff --git a/rtl8822cs/os_dep/linux/ioctl_cfg80211.c b/rtl8822cs/os_dep/linux/ioctl_cfg80211.c
index b58f76c2..f68e2af7 100644
--- a/rtl8822cs/os_dep/linux/ioctl_cfg80211.c
+++ b/rtl8822cs/os_dep/linux/ioctl_cfg80211.c
@@ -190,6 +190,9 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, u8 ch, u8 bw, u8 offset,
 {
 	struct wiphy *wiphy = adapter_to_wiphy(adapter);
 	u8 ret = _SUCCESS;
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 2))
+	unsigned int link_id = 0;
+#endif
 
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 8, 0))
 	struct cfg80211_chan_def chdef;
@@ -204,7 +207,9 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, u8 ch, u8 bw, u8 offset,
 
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0))
 	if (started) {
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0)) || defined(CONFIG_MLD_KERNEL_PATCH)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0))
+		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, link_id, 0, false);
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0)) || defined(CONFIG_MLD_KERNEL_PATCH)
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 15, 144))
 		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0, 0, false, punct_bitmap);
 #else
@@ -233,7 +238,9 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, u8 ch, u8 bw, u8 offset,
 	if (!rtw_cfg80211_allow_ch_switch_notify(adapter))
 		goto exit;
 
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0)) || defined(CONFIG_MLD_KERNEL_PATCH)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0))
+	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, link_id);
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0)) || defined(CONFIG_MLD_KERNEL_PATCH)
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 15, 144))
 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, 0, punct_bitmap);
 #else
@@ -5499,10 +5506,20 @@ static int rtw_cfg80211_set_beacon_ies(struct net_device *net, const u8 *head,
 }
 
 static int cfg80211_rtw_change_beacon(struct wiphy *wiphy, struct net_device *ndev,
-		struct cfg80211_beacon_data *info)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 7, 0))
+				      struct cfg80211_ap_update *param
+#else
+				      struct cfg80211_beacon_data *param
+#endif
+)
 {
 	int ret = 0;
 	_adapter *adapter = (_adapter *)rtw_netdev_priv(ndev);
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 7, 0))
+	struct cfg80211_beacon_data *info = &param->beacon;
+#else
+	struct cfg80211_beacon_data *info = param;
+#endif
 
 	RTW_INFO(FUNC_NDEV_FMT"\n", FUNC_NDEV_ARG(ndev));
 
diff --git a/rtl8822cs/os_dep/linux/os_intfs.c b/rtl8822cs/os_dep/linux/os_intfs.c
index b0104810..2652670a 100644
--- a/rtl8822cs/os_dep/linux/os_intfs.c
+++ b/rtl8822cs/os_dep/linux/os_intfs.c
@@ -22,6 +22,10 @@ MODULE_DESCRIPTION("Realtek Wireless Lan Driver");
 MODULE_AUTHOR("Realtek Semiconductor Corp.");
 MODULE_VERSION(DRIVERVERSION);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 8, 0)
+#define strlcpy(p, q, s) strscpy(p, q, s)
+#endif
+
 /* module param defaults */
 int rtw_chip_version = 0x00;
 int rtw_rfintfs = HWPI;
diff --git a/rtl8822cs/os_dep/linux/wifi_regd.c b/rtl8822cs/os_dep/linux/wifi_regd.c
index 5ac4b044..faa9e59b 100644
--- a/rtl8822cs/os_dep/linux/wifi_regd.c
+++ b/rtl8822cs/os_dep/linux/wifi_regd.c
@@ -154,7 +154,10 @@ static void rtw_regd_schedule_dfs_chan_update(struct wiphy *wiphy)
 			return;
 		rtw_regd_set_du_chdef(wiphy);
 	}
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0)) || defined(CONFIG_MLD_KERNEL_PATCH)
+
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0))
+	cfg80211_ch_switch_notify(wiphy_data->du_wdev->netdev, &wiphy_data->du_chdef, 0);
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0)) || defined(CONFIG_MLD_KERNEL_PATCH)
 #if (LINUX_VERSION_CODE > KERNEL_VERSION(5, 15, 144))
 	cfg80211_ch_switch_notify(wiphy_data->du_wdev->netdev, &wiphy_data->du_chdef,
 			          link_id, 0);
@@ -968,6 +971,9 @@ static void async_cac_change_work_hdl(_workitem *work)
 	struct rtw_wiphy_data *wiphy_data = container_of(work, struct rtw_wiphy_data, async_cac_change_work);
 	struct async_cac_change_evt *evt;
 	_list *list, *head = &wiphy_data->async_cac_change_list;
+	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 12, 0))
+	unsigned int link_id = 0;
+	#endif
 
 	while (1) {
 		_rtw_mutex_lock_interruptible(&wiphy_data->async_cac_change_mutex);
@@ -982,7 +988,13 @@ static void async_cac_change_work_hdl(_workitem *work)
 		evt = LIST_CONTAINOR(list, struct async_cac_change_evt, list);
 
 		rtnl_lock();
+
+		#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 12, 0))
+		cfg80211_cac_event(evt->netdev, &evt->chandef, evt->event, GFP_KERNEL, link_id);
+		#else
 		cfg80211_cac_event(evt->netdev, &evt->chandef, evt->event, GFP_KERNEL);
+		#endif
+
 		rtnl_unlock();
 
 		rtw_mfree(evt, sizeof(*evt));
@@ -1048,6 +1060,9 @@ static void rtw_cfg80211_cac_event(struct rf_ctl_t *rfctl, u8 band_idx
 	_adapter *iface;
 	int i;
 	bool async;
+	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 12, 0))
+	unsigned int link_id = 0;
+	#endif
 
 	if (!ifbmp) {
 		RTW_WARN(FUNC_HWBAND_FMT" ifbmp is empty", caller, HWBAND_ARG(band_idx));
@@ -1073,7 +1088,11 @@ static void rtw_cfg80211_cac_event(struct rf_ctl_t *rfctl, u8 band_idx
 		if (async)
 			cfg80211_cac_event_async(iface->pnetdev, &chdef, event);
 		else
+		#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 12, 0))
+			cfg80211_cac_event(iface->pnetdev, &chdef, event, GFP_KERNEL, link_id);
+		#else
 			cfg80211_cac_event(iface->pnetdev, &chdef, event, GFP_KERNEL);
+		#endif
 	}
 }
 
@@ -1102,6 +1121,7 @@ void rtw_cfg80211_cac_finished_event(struct rf_ctl_t *rfctl, u8 band_idx
 	struct wiphy *wiphy = dvobj_to_wiphy(dvobj); /* TODO: hwband specific wiphy mapping */
 	_adapter *iface;
 	int i;
+	bool cac;
 
 	if (!wiphy_ext_feature_isset(wiphy, NL80211_EXT_FEATURE_DFS_OFFLOAD))
 		return;
@@ -1110,8 +1130,15 @@ void rtw_cfg80211_cac_finished_event(struct rf_ctl_t *rfctl, u8 band_idx
 		iface = dvobj->padapters[i];
 		if (!iface || !(ifbmp & BIT(iface->iface_id)))
 			continue;
+
+		#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 12, 0))
+		cac = iface->rtw_wdev->links[0].cac_started;
+		#else
+		cac = iface->rtw_wdev->cac_started;
+		#endif
+
 		/* finish only for wdev with cac_started */
-		if (!iface->rtw_wdev || !iface->rtw_wdev->cac_started)
+		if (!iface->rtw_wdev || !cac)
 			ifbmp &= ~BIT(iface->iface_id);
 	}
 
@@ -1125,6 +1152,7 @@ void rtw_cfg80211_cac_aborted_event(struct rf_ctl_t *rfctl, u8 band_idx
 	struct wiphy *wiphy = dvobj_to_wiphy(dvobj); /* TODO: hwband specific wiphy mapping */
 	_adapter *iface;
 	int i;
+	bool cac;
 
 	if (!wiphy_ext_feature_isset(wiphy, NL80211_EXT_FEATURE_DFS_OFFLOAD))
 		return;
@@ -1133,8 +1161,14 @@ void rtw_cfg80211_cac_aborted_event(struct rf_ctl_t *rfctl, u8 band_idx
 		iface = dvobj->padapters[i];
 		if (!iface || !(ifbmp & BIT(iface->iface_id)))
 			continue;
+
+		#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 12, 0))
+		cac = iface->rtw_wdev->links[0].cac_started;
+		#else
+		cac = iface->rtw_wdev->cac_started;
+		#endif
 		/* abort only for wdev with cac_started */
-		if (!iface->rtw_wdev || !iface->rtw_wdev->cac_started)
+		if (!iface->rtw_wdev || !cac)
 			ifbmp &= ~BIT(iface->iface_id);
 	}
 
@@ -1167,6 +1201,7 @@ void rtw_cfg80211_cac_force_finished(struct rf_ctl_t *rfctl, u8 band_idx
 	struct ieee80211_channel *chan;
 	bool need_start = false;
 	u8 finished_ifbmp, started_ifbmp;
+	bool cac;
 
 	if (!wiphy_ext_feature_isset(wiphy, NL80211_EXT_FEATURE_DFS_OFFLOAD))
 		return;
@@ -1221,9 +1256,14 @@ void rtw_cfg80211_cac_force_finished(struct rf_ctl_t *rfctl, u8 band_idx
 			started_ifbmp &= ~BIT(iface->iface_id);
 			continue;
 		}
-		if (need_start && iface->rtw_wdev->cac_started)
+		#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 12, 0))
+		cac = iface->rtw_wdev->links[0].cac_started;
+		#else
+		cac = iface->rtw_wdev->cac_started;
+		#endif
+		if (need_start && cac)
 			started_ifbmp &= ~BIT(iface->iface_id);
-		else if (!need_start && !iface->rtw_wdev->cac_started)
+		else if (!need_start && !cac)
 			finished_ifbmp &= ~BIT(iface->iface_id);
 	}
 
